plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java

```diff
@@ -13,28 +13,118 @@
  */
 package org.apache.kudu.client;
 
+import io.trino.plugin.kudu.KuduClientSession;
+import io.trino.plugin.kudu.KuduClientWrapper;
 import io.trino.spi.TrinoException;
 
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkState;
 import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
 import static java.lang.String.format;
+import static org.apache.kudu.client.SessionConfiguration.FlushMode.MANUAL_FLUSH;
 
 /**
- * Operation.getChangeType() is package private
+ * Not thread safe
+ * This class is used to buffer operations and apply them in a batch and verify the batch was successfully applied.
+ * This gives:
+ * - performance of not flushing after every operation
+ * - correctness of not flushing in the background and hoping operations succeed
+ * The buffer is flushed when:
+ * - The KuduOperationApplier is closed
+ * - Or if the KuduOperationApplier reaches the bufferMaxOperations
+ * Note: Operation.getChangeType() is package private
  */
 public final class KuduOperationApplier
+        implements AutoCloseable
 {
-    private KuduOperationApplier()
+    private static final int bufferMaxOperations = 1000;
+
+    private int currentOperationsInBuffer;
+    private final KuduSession kuduSession;
+
+    private KuduOperationApplier(KuduSession kuduSession)
     {
+        kuduSession.setFlushMode(MANUAL_FLUSH);
```

Comment:
Can we configure the `kuduSession` to have manual_flush as its default flush mode.

Reply:
It is tricky to do without introducing extra abstraction in `ForwardingKuduClient`:
https://github.com/trinodb/trino/blob/master/plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/ForwardingKuduClient.java#L98-L101

And we guard against the case of incorrectly using the wrong flush mode here. I'm inclined to leave the configured default flush mode.